### PR TITLE
Add TraitThingIFAPIBuilder and part ofTraitThingIFAPI

### PR DESCRIPTION
**Pull request by yongpingchen**
- Add `TraitThingIFAPI` for using trait.
- Add some methods of `ThingIFAPI` to `TraitThingIFAPI`. These methods has the same implementation, so just call the same method of ThingIFAPI instance. 
   Refer https://github.com/KiiPlatform/thing-if-iOSSDK/pull/244
- Add Builder to create instance of `TraitThingIFAPI` 

Out of scope
- command methods in `TraitThingIFAPI`
- trigger methods in `TraitThingIFAPI`
- state methods in `TraitThingIFAPI`
- not tests (since it is in design phase)

### CODE REVIEW

File: thingif/src/main/java/com/kii/thingif/TraitThingIFAPIBuilder.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+public class TraitThingIFAPIBuilder {
+
+    private static final String TAG = ThingIFAPIBuilder.class.getSimpleName();
+    private final Context context;
+    private final KiiApp app;
+    private final Owner owner;
+    private Target target;
+    private String installationID;
+    private String tag;
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/TraitThingIFAPIBuilder.java]: I think adding Nullable/NonNull to fields is good mannter.

File: thingif/src/main/java/com/kii/thingif/TraitThingIFAPIBuilder.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    /**
+     * Instantiate new ThingIFAPIBuilder without Context.
+     * This method is for internal use only. Do not call it from your application.
+     *
+     * @param app Kii Cloud Application.
+     * @param owner Specify who uses the ThingIFAPI.
+     * @return ThingIFAPIBuilder instance.
+     */
+    @NonNull
+    public static TraitThingIFAPIBuilder _newBuilder(
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/TraitThingIFAPIBuilder.java]: Why is this method required?

File: thingif/src/main/java/com/kii/thingif/TraitThingIFAPIBuilder.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        this.tag = tag;
+        return this;
+    }
+
+    /**
+     * Set InstallationID to the ThingIFAPI.
+     * @param installationID installation id
+     * @return builder instance for chaining call.
+     */
+    public TraitThingIFAPIBuilder setInstallationID(String installationID) {
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/TraitThingIFAPIBuilder.java]: Please add Nonnull/Nullable annotation.

File: thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java [hunk 4]
```diff
@@ -0,0 +1,465 @@
...
+     * <BR>
+     * For details please refer to the {@link #loadFromStoredInstance(Context)} document.
+     *
+     * @param context context
+     * @param  tag specified when the ThingIFAPI has been built.
+     * @return ThingIFAPI instance.
+     * @throws StoredThingIFAPIInstanceNotFoundException when the instance has not stored yet.
+     */
+    @NonNull
+    public static TraitThingIFAPI loadFromStoredInstance(@NonNull Context context, String tag) throws StoredThingIFAPIInstanceNotFoundException {
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java]: Please add annotation to tag.

File: thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java [hunk 5]
```diff
@@ -0,0 +1,465 @@
...
+     *
+     * @param tag tag to specify stored instance.
+     */
+    public static void removeStoredInstance(@Nullable String tag) {
+        SharedPreferences preferences = getSharedPreferences();
+        SharedPreferences.Editor editor = preferences.edit();
+        editor.remove(getSharedPreferencesKey(tag));
+        editor.apply();
+    }
+    private static void saveInstance(TraitThingIFAPI instance) {
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java]: Please add annotation to instance.

File: thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java [hunk 6]
```diff
@@ -0,0 +1,465 @@
...
+        SharedPreferences preferences = getSharedPreferences();
+        if (preferences != null) {
+            SharedPreferences.Editor editor = preferences.edit();
+            editor.putString(getSharedPreferencesKey(instance.tag), GsonRepository.gson().toJson(instance));
+            editor.apply();
+        }
+    }
+    private static String getSharedPreferencesKey(String tag) {
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java]: Please add annotation to tag.

File: thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java [hunk 7]
```diff
@@ -0,0 +1,465 @@
...
+        return SHARED_PREFERENCES_KEY_INSTANCE + (tag == null ? "" : "_"  +tag);
+    }
+
+    TraitThingIFAPI(
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java]: Why this method is required?

File: thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java [hunk 8]
```diff
@@ -0,0 +1,465 @@
...
+     * @throws ThingIFRestException Thrown when server returns error response.
+     */
+    @NonNull
+    @WorkerThread
+    public Target onboard(
+            @NonNull String vendorThingID,
+            @NonNull String thingPassword,
+            @Nullable OnboardWithVendorThingIDOptions options)
+            throws ThingIFException {
+        return thingIfApi.onboard(vendorThingID, thingPassword, options);
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java]: Adding `this` to a field is good manner.

File: thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java [hunk 9]
```diff
@@ -0,0 +1,465 @@
...
+     * @throws ThingIFException Thrown when failed to connect IoT Cloud Server.
+     * @throws ThingIFRestException Thrown when server returns error response.
+     */
+    @NonNull
+    @WorkerThread
+    public Target onboard(
+            @NonNull String thingID,
+            @NonNull String thingPassword) throws
+            ThingIFException {
+        return thingIfApi.onboard(thingID, thingPassword);
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java]: Adding `this` to a field is good manner.

File: thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java [hunk 10]
```diff
@@ -0,0 +1,465 @@
...
+     * @throws ThingIFRestException Thrown when server returns error response.
+     */
+    @NonNull
+    @WorkerThread
+    public Target onboard(
+            @NonNull String thingID,
+            @NonNull String thingPassword,
+            @Nullable OnboardWithThingIDOptions options)
+            throws ThingIFException {
+        return thingIfApi.onboard(thingID, thingPassword, options);
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java]: Adding `this` to a field is good manner.

File: thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java [hunk 11]
```diff
@@ -0,0 +1,465 @@
...
+     * @return Target instance can be used to operate target, manage resources of the target.
+     * @throws IllegalStateException Thrown when this instance is already onboarded.
+     * @throws ThingIFException Thrown when failed to connect IoT Cloud Server.
+     * @throws ThingIFRestException Thrown when server returns error response.
+     */
+    public EndNode onboardEndnodeWithGateway(
+            @NonNull PendingEndNode pendingEndNode,
+            @NonNull String endnodePassword)
+            throws ThingIFException {
+        return thingIfApi.onboardEndnodeWithGateway(pendingEndNode, endnodePassword);
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java]: Adding `this` to a field is good manner.

File: thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java [hunk 12]
```diff
@@ -0,0 +1,465 @@
...
+     * @throws IllegalStateException Thrown when this instance is already onboarded.
+     * @throws ThingIFException Thrown when failed to connect IoT Cloud Server.
+     * @throws ThingIFRestException Thrown when server returns error response.
+     */
+    public EndNode onboardEndnodeWithGateway(
+            @NonNull PendingEndNode pendingEndNode,
+            @NonNull String endnodePassword,
+            @Nullable OnboardEndnodeWithGatewayOptions options)
+            throws ThingIFException {
+        return thingIfApi.onboardEndnodeWithGateway(pendingEndNode, endnodePassword, options);
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java]: Adding `this` to a field is good manner.

File: thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java [hunk 13]
```diff
@@ -0,0 +1,465 @@
...
+    }
+
+    /**
+     * Checks whether on boarding is done.
+     * @return true if done, otherwise false.
+     */
+    public boolean onboarded()
+    {
+        return thingIfApi.onboarded();
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java]: Adding `this` to a field is good manner.

File: thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java [hunk 14]
```diff
@@ -0,0 +1,465 @@
...
+     * @throws ThingIFRestException Thrown when server returns error response.
+     * @see #installPush(String, PushBackend, boolean) for development/production environment installation.
+     */
+    @NonNull
+    @WorkerThread
+    public String installPush(
+            @Nullable String deviceToken,
+            @NonNull PushBackend pushBackend
+    ) throws ThingIFException {
+        return thingIfApi.installPush(deviceToken, pushBackend);
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java]: Adding `this` to a field is good manner.

File: thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java [hunk 15]
```diff
@@ -0,0 +1,465 @@
...
+     * @throws ThingIFRestException Thrown when server returns error response.
+     */
+    @NonNull
+    @WorkerThread
+    public String installPush(
+            @Nullable String deviceToken,
+            @NonNull PushBackend pushBackend,
+            boolean development
+    ) throws ThingIFException{
+        return thingIfApi.installPush(deviceToken, pushBackend, development);
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java]: Adding `this` to a field is good manner.

File: thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java [hunk 16]
```diff
@@ -0,0 +1,465 @@
...
+    }
+    /**
+     * Get installationID if the push is already installed.
+     * null will be returned if the push installation has not been done.
+     * @return Installation ID used in IoT Cloud.
+     */
+    @Nullable
+    public String getInstallationID() {
+        return thingIfApi.getInstallationID();
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java]: Adding `this` to a field is good manner.

File: thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java [hunk 17]
```diff
@@ -0,0 +1,465 @@
...
+     *                       {@link #installPush(String, PushBackend)}
+     *                       if null is specified, value obtained by
+     *                       {@link #getInstallationID()} is used.
+     * @throws ThingIFException Thrown when failed to connect IoT Cloud Server.
+     * @throws ThingIFRestException Thrown when server returns error response.
+     */
+    @NonNull
+    @WorkerThread
+    public void uninstallPush(@NonNull String installationID) throws ThingIFException {
+        thingIfApi.uninstallPush(installationID);
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java]: Adding `this` to a field is good manner.

File: thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java [hunk 18]
```diff
@@ -0,0 +1,465 @@
...
+    /**
+     * Get the Vendor Thing ID of specified Target.
+     *
+     * @return Vendor Thing ID
+     * @throws ThingIFException Thrown when failed to connect IoT Cloud Server.
+     */
+    @NonNull
+    @WorkerThread
+    public String getVendorThingID() throws ThingIFException {
+        return thingIfApi.getVendorThingID();
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java]: Adding `this` to a field is good manner.

File: thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java [hunk 19]
```diff
@@ -0,0 +1,465 @@
...
+    /**
+     * Update the Vendor Thing ID of specified Target.
+     *
+     * @param newVendorThingID New vendor thing id
+     * @param newPassword New password
+     * @throws ThingIFException Thrown when failed to connect IoT Cloud Server.
+     */
+    @WorkerThread
+    public void updateVendorThingID(@NonNull String newVendorThingID, @NonNull String newPassword) throws ThingIFException {
+        thingIfApi.updateVendorThingID(newVendorThingID, newPassword);
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java]: Adding `this` to a field is good manner.

File: thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java [hunk 20]
```diff
@@ -0,0 +1,465 @@
...
+    }
+
+    /** Get Kii App
+     * @return Kii Cloud Application.
+     */
+    @NonNull
+    public KiiApp getApp() {
+        return thingIfApi.getApp();
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java]: Adding `this` to a field is good manner.

File: thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java [hunk 21]
```diff
@@ -0,0 +1,465 @@
...
+    }
+    /**
+     * Get AppID
+     * @return app ID
+     */
+    @NonNull
+    public String getAppID() {
+        return thingIfApi.getAppID();
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java]: Adding `this` to a field is good manner.

File: thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java [hunk 22]
```diff
@@ -0,0 +1,465 @@
...
+    }
+    /**
+     * Get AppKey
+     * @return app key
+     */
+    @NonNull
+    public String getAppKey() {
+        return thingIfApi.getAppKey();
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java]: Adding `this` to a field is good manner.

File: thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java [hunk 23]
```diff
@@ -0,0 +1,465 @@
...
+    }
+    /**
+     * Get base URL
+     * @return base URL
+     */
+    @NonNull
+    public String getBaseUrl() {
+        return thingIfApi.getBaseUrl();
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java]: Adding `this` to a field is good manner.

File: thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java [hunk 24]
```diff
@@ -0,0 +1,465 @@
...
+    }
+    /**
+     * Get owner who uses the ThingIFAPI.
+     * @return owner
+     */
+    @NonNull
+    public Owner getOwner() {
+        return thingIfApi.getOwner();
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java]: Adding `this` to a field is good manner.

File: thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java [hunk 25]
```diff
@@ -0,0 +1,465 @@
...
+    }
+
+    /**
+     * Get target thing that is operated by the ThingIFAPI.
+     * @return target of this ThingIFAPI.
+     */
+    @Nullable
+    public Target getTarget() {
+        return thingIfApi.getTarget();
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java]: Adding `this` to a field is good manner.

File: thingif/src/main/java/com/kii/thingif/TraitThingIFAPIBuilder.java [hunk 26]
```diff
@@ -0,0 +1,131 @@
...
+
+import android.content.Context;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import com.kii.thingif.internal.utils._Log;
+
+import java.text.MessageFormat;
+
+public class TraitThingIFAPIBuilder {
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/TraitThingIFAPIBuilder.java]: This builder class should be inner class of `TraitThingIFAPI`.

File: thingif/src/main/java/com/kii/thingif/TraitThingIFAPIBuilder.java [hunk 27]
```diff
@@ -0,0 +1,131 @@
...
+        }
+        return new TraitThingIFAPIBuilder(null, app, owner);
+    }
+
+    /**
+     * Set target thing to the ThingIFAPI.
+     * @param target target of {@link ThingIFAPI} instance.
+     * @return builder instance for chaining call.
+     */
+    public TraitThingIFAPIBuilder setTarget(Target target) {
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/TraitThingIFAPIBuilder.java]: Please add Nullable/Nonnull annotation.

File: thingif/src/main/java/com/kii/thingif/TraitThingIFAPIBuilder.java (hunk 26 above)
Comment (yongpingchen) [inline thingif/src/main/java/com/kii/thingif/TraitThingIFAPIBuilder.java]: yes. Better to be inner class

File: thingif/src/main/java/com/kii/thingif/TraitThingIFAPIBuilder.java (hunk 2 above)
Comment (yongpingchen) [inline thingif/src/main/java/com/kii/thingif/TraitThingIFAPIBuilder.java]: I will remove it

File: thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java (hunk 7 above)
Comment (yongpingchen) [inline thingif/src/main/java/com/kii/thingif/TraitThingIFAPI.java]: will remove it after builder introduced as inner class

**Comment by yongpingchen** [comment]
@shinsuke-nara please confirm the commits

**Comment by shinsuke-nara** [comment]
@yongpingchen LGTM! Please merge after circle ci build.

**Pull request merged**